Add a GNOME Shell Extension

nathandyer opened the pull request:
## Status

Work in progress

## Description of Changes

Fixes # 6531

Changes proposed in this pull request:

This adds a GNOME Shell extension to the Journalist and Admin Workstations. The goal of the extension is to provide a more consistent and better-supported method for linking a user to their SecureDrop interfaces, and to make it easier to access the management features of SecureDrop.

As of now, the extension looks like this:

![sdgse](https://user-images.githubusercontent.com/6131358/209209498-80458c61-8488-4994-ab26-75908c0f1c4f.png)

The Shell Extension gets saved into the appropriate .dotfiles directory as part of the `tailsconfig` Ansible task, during install or upgrade.

Although the code here is functional and can be included in the main project as-is, the intent of this PR is to provide a firmer foundation for testing and discussion. Every menu entry in the extension is functional, and it is ready for wider testing.

There are some additional concerns that need to be ironed out:

* Do we want to keep the icon in the titlebar where the extension is accessed? If so, do we use the generic icon, or do we create/install a symbolic SD icon?
* Do we want to dynamically hide the administrative options when it is being used only on a journalist workstation? If yes, what is the most reliable method of detecting that difference?
* Do we want to remove the current desktop icons altogether, or have this in addition to those icons (either temporarily or indefinitely until that support is removed from Tails)? 

## Testing

This needs to be widely tested across multiple installs.

At a minimum, we should:
- [x] Test that the extension gets added and activated with a clean install
- [ ] Test that the extension gets added and activated while upgrading an existing install
- [x] Test that each menu item in the extension reliably triggers the labeled action

### Fresh install, ssh over Tor
- [ ] Fresh install using this branch on the Admin Workstation completes successfully
- [ ]  Menu is available after rebooting, with icon displayed correctly
- [ ] Journalist (interfaces, updater, file browser) and Admin (SSH to server) menu options are displayed and work as expected

### Fresh install, ssh over local
- [ ] Fresh install using this branch on the Admin Workstation completes successfully
- [ ]  Menu is available after rebooting, with icon displayed correctly
- [ ] Journalist (interfaces, updater, file browser) and Admin (SSH to server) menu options are displayed and work as expected

### Admin Upgrade (any ssh option)
- set up a prod (VM OK) install using the current latest version (2.5.2)
- On the Admin workstation, simulate running the GUI updater via these commands in sequence:
  ```
  cd ~/amnesia/Persistent/securedrop && git checkout gnome-shell-extension
  ./securedrop-admin setup
  ./securedrop-admin --force tailsconfig
  ```
- [ ] `tailsconfig` completes successfully, prompting user to reboot
- [ ]  Menu is available after rebooting, with icon displayed correctly
- [ ] Journalist (interfaces, updater, file browser) and Admin (SSH to server) menu options are displayed and work as expected


### Journalist workstation - fresh install
- [ ] Following JW setup instructions and rebooting results in menu being available
- [ ] Admin (ssh) options are *not* displayed, other options are and work as expected


### Journalist workstation - upgrade 
- On an existing Journalist workstation, set up with current release (2.5.2), simulate running the GUI updater via these commands in sequence:
  ```
  cd ~/amnesia/Persistent/securedrop && git checkout gnome-shell-extension
  ./securedrop-admin setup
  ./securedrop-admin --force tailsconfig
  ```
- [ ] `tailsconfig` completes successfully, prompting user to reboot
- [ ]  Menu is available after rebooting, with icon displayed correctly
- [ ] Journalist (interfaces, updater, file browser)  menu options are displayed and work as expected
- [ ] Admin (ssh) options are *not dispalyed



## Deployment

Deployment should be straight-forward, although as mentioned above, we need to test thoroughly for both clean install and upgrade paths. 

I have already tested with a clean install, and it _should_ work with an upgrade, but would appreciate other confirmations.

## Checklist

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made changes to the system configuration:

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you added or removed a file deployed with the application:

- [ ] I have updated AppArmor rules to include the change

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

Choose one of the following:

- [x] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation
- [ ] These changes do not require documentation

zekehuntergreen committed:
empty commit to rerun ci after setting up circle ci project for fork

nathandyer committed:
First stab at adding a GNOME Shell Extension, and the necessary logic for Ansible to install and enable it.

nathandyer committed:
Fix broken commands, have Tails load extension with network hook, and make sure Ansible actually runs the new task.

nathandyer committed:
Don't run gnome-extensions command with elevated privileges (without this, the task will fail)

nathandyer committed:
Remove trailing spaces to appease the linter

nathandyer committed:
Don't show admin-related menu items if it's on a journalist workstation.

nathandyer commented:
The extension now checks to see if it's on an Admin workstation or Journalist workstation, and shows the appropriate actions accordingly.

I believe this is officially ready for wider review and testing by the team.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/files/securedrop_init.py:
```diff
@@ -129,6 +129,9 @@
 if is_tails:
```
Not strictly related but I think we can remove this call, `gnome-shell-extension-tool` appears to be deprecated.

nathandyer replied:
Unless I'm mistaken, I do believe `gnome-extensions` (or its older alias, `gnome-shell-extension-tool`) is still in use, and is necessary here. There are instances, generally outside of Tails installations, where folks will have Shell extensions installed, but not enabled, and without this tool they don't have a way to turn them on and off dynamically while leaving them installed on the system (there are GUI tools like the new Extensions app and GNOME Tweaks that let you do this, but programmatically I believe it's still necessary to use `gnome-extensions`, because even if it's in the right place in the dotfiles, it needs the explicit enable action to launch).

zenmonkeykstop replied:
AIUI They're two different utilities, not an alias. You do need the gnome-extensions call just below, as that enables your new extension, but the one above is a legacy thing from Tails 4 where it was necessary to reload the desktop icons extension so that the icons would refresh after being enabled. That's the call this refers to.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/files/securedrop@securedrop.freedom.press/extension.js:
```diff
@@ -0,0 +1,129 @@
+        // Check for SSH keys - if they don't exist, this is a Journalist workstation
+        // (and therefore the admin options should remain hidden)
+        let is_admin = false;
+        let [ok, out, err, exit] = GLib.spawn_command_line_sync('/bin/bash -c "test -f ~/Persistent/securedrop/install_files/ansible-base/app-ssh.auth_private"');
```
This doesn't cover the (admittedly super-rare) case where the workstation is set up for SSH-over-local, as the `.auth_private` file will not exist.

(I don't think this is a requirement for release, as a: SSH-over-local is bad and wrong in the vast majority of cases and b: said users can still SSH in as before.)

nathandyer replied:
I changed this to look for `tor_v3_keys.json` rather than `app-ssh.auth_private`, which should cover the case where a workstation is set up for SSH-over-local.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/tasks/install_shell_extension.yml:
```diff
@@ -0,0 +1,37 @@
+- name: Create the SecureDrop GNOME Shell Extension directory
+  become: yes
+  file:
+    path: "{{ tails_config_live_dotfiles }}/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press"
```
Files here are symlinked into the home directory at the start of a session - so this directory and associated files are not available when the `gnome-extensions enable` command runs below. It looks like the desktop icon tasks work around this by initially installing the icons both into `~amnesia/.local/share/applications` and into `/live/persistence/TailsData_unlocked/dotfiles/.local/share/applications` so that they're available both on the first run and on subsequent sessions. Something similar might work here.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/files/securedrop@securedrop.freedom.press/extension.js:
```diff
@@ -0,0 +1,129 @@
+        let journalist = new PopupMenu.PopupMenuItem('Launch Journalist Interface');
+        journalist.connect('activate', () => {
+       	    let [ok, out, err, exit] = GLib.spawn_command_line_sync('/bin/bash -c "cut -d: -f1 ~/Persistent/securedrop/install_files/ansible-base/app-journalist.auth_private"');
+       	    if (out.length > 0) {
```
Can use precise lengths here.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/files/securedrop@securedrop.freedom.press/extension.js:
```diff
@@ -0,0 +1,129 @@
+        let source = new PopupMenu.PopupMenuItem('Launch Source Interface');
+        source.connect('activate', () => {
+            let [ok, out, err, exit] = GLib.spawn_command_line_sync('/bin/bash -c "cat ~/Persistent/securedrop/install_files/ansible-base/app-sourcev3-ths"');
+       	    if (out.length > 0) {
```
Can use precise lengths here and validate that it ends in .onion

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/files/securedrop@securedrop.freedom.press/extension.js:
```diff
@@ -0,0 +1,129 @@
+            y_align: Clutter.ActorAlign.START,
+        });
+        
+        let updates = new PopupMenu.PopupMenuItem('Check for SecureDrop Updates');
```
The updater action could also be exposed on Journalist workstations.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/files/securedrop@securedrop.freedom.press/extension.js:
```diff
@@ -0,0 +1,129 @@
+        
+        let app_server_ssh = new PopupMenu.PopupMenuItem('SSH into the App Server');
+        app_server_ssh.connect('activate', () => {
+            Util.trySpawnCommandLine(`gnome-terminal -- ssh app`);
```
It's not guaranteed that the server alias is 'app', the value set in `sdconfig` is used in the .ssh/config file.

Update: one way to account for this would be to make these files templates instead and use Ansible to populate the site-specific information. It would also solve the problem with detecting ssh settings and onion addresses - rather than inferring them at runtime by looking at the auth_private files on the workstation, bake them into the extension as it's being installed.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/files/securedrop@securedrop.freedom.press/extension.js:
```diff
@@ -0,0 +1,129 @@
+        
+        let mon_server_ssh = new PopupMenu.PopupMenuItem('SSH into the Monitor Server');
+        mon_server_ssh.connect('activate', () => {
+            Util.trySpawnCommandLine(`gnome-terminal -- ssh mon`);
```
It's not guaranteed that the server alias is 'mon', the value set in sdconfig is used in the .ssh/config file.

zenmonkeykstop reviewed (changes requested):
Tested upgrade path  on 5.8 from current version (2.5.1) of SD, by:
- checking out this branch
- running `./securedrop-admin tailsconfig`

The command fails when running the `gnome-extensions enable` task, as the files installed in the persistent volume have not been symlinked into  the `amnesia` home directory. After rebooting, the symlink is set up and `tailsconfig` succeeds.

nathandyer commented:
Thanks so much for the review, @zenmonkeykstop! I plan to step through these issues and have an updated version ready to test before the end of the week.

nathandyer committed:
Switch to checking for tor_v3_keys.json to determine between Journalist and Admin workstations.

nathandyer committed:
Use explicit length checks and validate the URL contains .onion

nathandyer committed:
Show update option, even on Journalist workstation.

nathandyer committed:
Copy files to running system in addition to Persistent volume

nathandyer committed:
Load app and mon hostnames dynamically

nathandyer committed:
Fix a couple bugs recently introduced that prevent the extension from loading.

nathandyer committed:
Switch to using an Ansible template.

cfm commented:
@nathandyer and I met yesterday to discuss how this extension can be localized.  Here are our conclusions for the record.  @nathandyer, please comment or edit if I've missed or mischaracterized anything!

## What we learned

For the current desktop shortcuts, e.g. `desktop-source-icon`:

1. `desktop-source-icon.j2.in` is untranslated and unconfigured, until `i18n_tool.py` turns it into...
2. `desktop-source-icon.j2`, which contains all translations but is still configured, until Ansible turns it into...
3. `desktop-source-icon`, which contains all translations and is configured for the instance.

We could preserve this structure for `extension.js`, except that `i18n_tool.py` is only able to act on `desktop-source-icon.j2.in` because `msgfmt` takes a `--desktop` flag.

## What we decided

Instead, @nathandyer will take advantage of the GNOME Shell's [dynamic support] for translations at runtime, just as the Source and Journalist Applications do.  The steps here are for @nathandyer to:

1. [x] Wrap strings in `gettext()`/`_()`
2. [ ] Remove ~`msgfmt` from~ `i18n_tool.py --translate-desktop`
3. [ ] Change  `i18n_tool.py` to `xgettext`-extract from `extension.js` (from step 1) ~and download in an extension-appropriate place (for step 2)~
    - https://github.com/freedomofpress/securedrop/blob/c2f11efb11b59041137ca73fd19b3af96da6c109/securedrop/i18n_tool.py#L239
4. [ ] Ansible installs `.po` catalogs per <https://gjs.guide/extensions/development/translations.html#preparing-an-extension>

I'm happy to review and/or pair on any of the above.  In addition, I'll look into:

5. [ ] In Weblate, should this be a new `securedrop/gnome-shell-extension` component or folded into `securedrop/securedrop`?  (We'll need to deprecate the `securedrop/desktop` component and update developers.securedrop.org accordingly.)

[dynamic support]: https://gjs.guide/extensions/development/translations.html#preparing-an-extension

cfm commented:
<https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1412619641>:

> 5. In Weblate, should this be a new `securedrop/gnome-shell-extension` component or folded into `securedrop/securedrop`?  (We'll need to deprecate the `securedrop/desktop` component and update developers.securedrop.org accordingly.)

I've traced the introduction of the `securedrop/desktop` component to 5e8dfb0d34b06699d234c5a7f74f4b1dd6a125de, which corroborates what we inferred yesterday:

https://github.com/freedomofpress/securedrop/blob/5e8dfb0d34b06699d234c5a7f74f4b1dd6a125de/docs/development/i18n.rst?plain=1#L127-L131

Since this is a constraint of the desktop icon that doesn't hold for the GNOME Shell extension, the latter can indeed use the `securedrop/securedrop` component directly.  I've updated the steps in <https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1412619641> to reflect that `i18n_tool.py --translate-desktop` can be removed altogether.  Documentation implications filed in freedomofpress/securedrop-dev-docs#34.

zenmonkeykstop commented:
One point of caution - I'm in agreement with removing the desktop icons in favor of the menu, but just in case we don't go that way:
- the removal should be easily revertible (preferably a single commit or group thereof)
- we shouldn't drop the Weblate config until this is approved and merged

(apologies if this is obvious)

edit: looks like the first bullet point is covered, and poetically at that :) it would be cool to preserve that commit if/when there's any squash/rebase action.

nathandyer commented:
Thanks @zenmonkeykstop, 100%, the plan is to confine the changes to i18n_tool to a single commit, so we can just remove or modify it as needed based on the decision surrounding the removal of desktop icons. We also won't make any changes to Weblate until (if) this is merged.

@cfm I just pushed commit [1cfd35b](https://github.com/freedomofpress/securedrop/pull/6712/commits/1cfd35b5b2a2551e6d8f844968e92425959f3d56) which is a first attempt at making those changes. I ran `make translate`, and it successfully created the new .pot file, which is a good sign. That said, I'm sure changes will be needed. If you wouldn't mind, please take a look if you have some time, and I'll make any necessary changes (and re-base to put them in a single commit once things are how we want them).

cfm reviewed (changes requested):
@nathandyer, I'm sorry I missed this during other reviews on Monday!  Your changes here look good, and the rename-with-additions of `desktop.pot` to `extension.pot` is very elegant, *if* the goal here is to use a separate `securedrop/extension` translation component in Weblate.  However, per <https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1413001898>, I don't see any reason not to fold the extension's translation into the main `securedrop/securedrop` component.  In that case, we want `i18n_tool.py --translate` to extract source strings from `extension.js` in addition to the Flask apps and templates, removing `i18n_tool.py --translate-{desktop,extension}` altogether.  Then `extension.js` can be installed with the main `securedrop/translations/` catalogs and load them at runtime, just like the Flask apps do (and as the desktop shortcut cannot).

Does this clarification from <https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1412619641> make sense?  Let me know if you'd like to chat further.

nathandyer commented:
Thank you @cfm, that makes complete sense. I'll take another swing at this and ping you when it's ready for another review.

nathandyer committed:
Use a SecureDrop icon in the Shell extension

nathandyer committed:
Add a button to open a file browser.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/templates/extension.js.in:
```diff
@@ -91,6 +91,11 @@ class Indicator extends PanelMenu.Button {
             Util.trySpawnCommandLine(`keepassxc`);
         });
+
+        let filebrowser = new PopupMenu.PopupMenuItem(_('Open File Browser'));
```
Is this overkill? there's already a Places menu

nathandyer replied:
@zenmonkeykstop One of the issues I've observed in various trainings and in discussions with the DST team is that figuring out how to even open the file browser can be challenging for new users. They don't associate the Places menu with browsing files, so they go to the Activities menu. Sometimes they can get to Nautilus quickly by searching for 'Files', but often they'll manually hunt around the various categories until they find an icon with a folder on it. By putting this option in the extension, it eliminates the confusion from Places, and the challenge of hunting around the Activities view for it.

zenmonkeykstop replied:
OK, but it also introduces a non-SD related menu item to a menu labelled SecureDrop, and as currently implemented it opens the Persistent directory, when we recommend that journalists default to using non-persistent directories instead (see https://docs.securedrop.org/en/stable/journalist.html#step-1-download-the-encrypted-submission). I don't know if we recommend use of the Persistent directory (`Persistent/Tor Browser`, yes) anywhere in the docs for submissions. There's also further opportunity for new user confusion here, as the first thing they'll see on using this option is a directory called `securedrop`, with which they should definitely not be messing.

The original intent of the PR as I understood it was to get our UX additions more in line with the Tails UX. This feels kindof counter to that intent (we're now working around perceived issues with it?) and a bit scope-creepy.

zenmonkeykstop replied:
This is still in place but not a blocker for merge.

nathandyer committed:
Make symbolic icon white instead of black so it's visible in the menu.

nathandyer committed:
Enable translations for the SecureDrop Extension

nathandyer committed:
Fix the icon appearance, set the position in the menu, and fix a bug with a button opening the wrong interface.

nathandyer committed:
Appease the linter.

nathandyer commented:
@cfm I believe the translation-related aspects are ready for review, whenever you have a moment to take a look. Thank you!

cfm reviewed (changes requested):
@nathandyer, I've taken a look at your new changes to `i18n_tool.py`, which do indeed extract source strings from the new `extension.js.in`.  Here's a patch which:
1. clarifies `LOCALE_DIR`'s mapping of Weblate components and translation directories (versus sources); and
2. makes sure to still extract source strings from everything else too.  :-)

Comparing the diff of `make translate` with and without this patch will demonstrate what I mean.  I'll note that (1) is genuinely confusing: the `desktop`  component extracts source strings *from* and writes translations *to* the same `install_files/ansible-base/roles/tails-config/templates` directory.  So it's a misleading example for what you're trying to do here in merging the extension's translations into the main `securedrop` component.

Let me know what you think—or if you'd like to look at this change together.

```patch
diff --git a/securedrop/i18n_tool.py b/securedrop/i18n_tool.py
index 60235052a..629329987 100755
--- a/securedrop/i18n_tool.py
+++ b/securedrop/i18n_tool.py
@@ -25,9 +25,10 @@ I18N_CONF = os.path.join(os.path.dirname(__file__), "i18n.json")
 # Map components of the "securedrop" Weblate project (keys) to their filesystem
 # paths (values) relative to the repository root.
 LOCALE_DIR = {
+    # The Flask apps and GNOME Shell extension share the "securedrop" component:
     "securedrop": "securedrop/translations",
+    # The legacy desktop shortcuts have their own:
     "desktop": "install_files/ansible-base/roles/tails-config/templates",
-    "extension": "install_files/ansible-base/roles/tails-config/templates",
 }
 
 
@@ -237,8 +238,14 @@ class I18NTool:
             "translate-messages", help=("Update and compile " "source and template translations")
         )
         translations_dir = join(dirname(realpath(__file__)), "translations")
-        extension_location = join(dirname(realpath(__file__)), "..", LOCALE_DIR["extension"])
-        sources = join(".,source_templates,journalist_templates,", extension_location)
+        sources = ",".join(
+            [
+                ".",
+                "source_templates",
+                "journalist_templates",
+                "../install_files/ansible-base/roles/tails-config/templates",
+            ]
+        )
         self.set_translate_parser(parser, translations_dir, sources)
         mapping = "babel.cfg"
         parser.add_argument(
```

nathandyer commented:
Thanks @cfm! Your explanation for your patch does indeed make sense to me, and I think it is a good and necessary change.

sssoleileraaa commented:
[Backlog pruning, 5/15] This is a keeper for the 2.6.0 release since a lot of work has already been done and several folks have tested this + is a Tails-only change. It's fairly easy to do comprehensive testing.

cfm commented on install_files/ansible-base/roles/tails-config/tasks/install_shell_extension.yml:
```diff
@@ -0,0 +1,172 @@
+  when: item.startswith(tails_config_amnesia_home)
+  with_items: "{{ tails_config_extension_directories }}"
+
+- name: Copy the extension metadata to the extension directory in Persistent Storage
+  become: yes
+  copy:
+    src: securedrop@securedrop.freedom.press/metadata.json
+    dest: "{{ tails_config_live_dotfiles }}/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/"
+    owner: amnesia
+    group: amnesia
+
+- name: Copy the extension metadata to the extension directory in the running system
+  become: yes
+  copy:
+    src: securedrop@securedrop.freedom.press/metadata.json
+    dest: "/home/amnesia/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/"
+    owner: amnesia
+    group: amnesia
```
Perhaps related to this failure mode:  Can this be DRY'd at all, by (1) installing the `.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press` tree into either the persistent storage or the running system and then (2) replicating it from the one to the other?

nathandyer replied:
Thanks @cfm! Off the top of my head, I don't think copying it into the tree and replicating it there buys us much operationally (you're still having essentially the same number of copy operations, with only the source changing). That said, I think I could clean this up a bit by using the same sort of logic that is currently in use by the Desktop Icons task, which loops over the install directories. I'll take a look at that change later today.

cfm reviewed (changes requested):
The extension (a) still requires a reboot during installation and (b) does not appear to activate on Tails 5.13.

First attempt:

* [ ] `securedrop-admin --force tailsconfig` succeeds.

```sh-session
amnesia@amnesia:~/Persistent/securedrop$ ./securedrop-admin update
[...]
HEAD is now at ea83806ce SecureDrop 2.5.2
INFO: Updated to SecureDrop 2.5.2.
amnesia@amnesia:~/Persistent/securedrop$ git checkout gnome-shell-extension
Previous HEAD position was ea83806ce SecureDrop 2.5.2
Switched to branch 'gnome-shell-extension'
Your branch is up to date with 'origin/gnome-shell-extension'.
amnesia@amnesia:~/Persistent/securedrop$ ./securedrop-admin --force tailsconfig
[...]
TASK [tails-config : Check for v3 Source Interface file] ***********************
ok: [localhost]

TASK [tails-config : Check for v3 Journalist Interface file] *******************
ok: [localhost]

TASK [tails-config : Check for site specific file] *****************************
ok: [localhost]

TASK [tails-config : Look up v3 Source Interface URL.] *************************
ok: [localhost]

TASK [tails-config : Look up v3 Journalist Interface URL.] *********************
ok: [localhost]

TASK [tails-config : Look up app server hostname] ******************************
ok: [localhost]

TASK [tails-config : Look up mon server hostname] ******************************
ok: [localhost]

TASK [tails-config : Create the SecureDrop GNOME Shell Extension directories] ***
changed: [localhost] => (item=/home/amnesia/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/)
changed: [localhost] => (item=/live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/)

TASK [tails-config : Set normal user ownership on subset of directories.] ******
ok: [localhost] => (item=/home/amnesia/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/)

TASK [tails-config : Copy the extension metadata to the extension directory in Persistent Storage] ***
changed: [localhost]

TASK [tails-config : Copy the extension metadata to the extension directory in the running system] ***
changed: [localhost]

TASK [tails-config : Copy the extension metadata to the extension directory in Persistent Storage] ***
ok: [localhost]

TASK [tails-config : Copy the extension metadata to the extension directory in the running system] ***
ok: [localhost]

TASK [tails-config : Copy the extension CSS to the extension directory in Persistent Storage] ***
changed: [localhost]

TASK [tails-config : Copy the extension CSS to the extension directory in the running system] ***
changed: [localhost]

TASK [tails-config : Copy the symbolic icon used for the shell extension in Persistent Storage] ***
changed: [localhost]

TASK [tails-config : Copy the symbolic icon used for the shell extension in the running system] ***
changed: [localhost]

TASK [tails-config : Set the right variable for source] ************************
ok: [localhost]

TASK [tails-config : Set the right variable for journalist] ********************
ok: [localhost]

TASK [tails-config : Set the right variable for app server hostname] ***********
ok: [localhost]

TASK [tails-config : Set the right variable for app server hostname] ***********
ok: [localhost]

TASK [tails-config : Assemble interface information for extension] *************
ok: [localhost]

TASK [tails-config : Create SecureDrop extension] ******************************
changed: [localhost] => (item=[{'src': 'extension.js.in', 'filename': 'extension.js', 'source_interface_address': 'fu4cp2ynauxf36fdtqluyis75gn33xs72p5bdfu7ve2kut4y7gbqvxid.onion', 'journalist_interface_address': '77aml4ggknlndjnkshytyfstwuywlpuncmvr6qnxyuszmzkocobsi4ad.onion', 'app_hostname': 'app', 'mon_hostname': 'mon'}, '/home/amnesia/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/'])
changed: [localhost] => (item=[{'src': 'extension.js.in', 'filename': 'extension.js', 'source_interface_address': 'fu4cp2ynauxf36fdtqluyis75gn33xs72p5bdfu7ve2kut4y7gbqvxid.onion', 'journalist_interface_address': '77aml4ggknlndjnkshytyfstwuywlpuncmvr6qnxyuszmzkocobsi4ad.onion', 'app_hostname': 'app', 'mon_hostname': 'mon'}, '/live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/'])

TASK [tails-config : Add extension translations in Persistent Storage] *********
changed: [localhost]

TASK [tails-config : Add extension translations in the running system] *********
changed: [localhost]

TASK [tails-config : Enable the new GNOME Shell extension] *********************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["gnome-extensions", "enable", "securedrop@securedrop.freedom.press"], "delta": "0:00:00.409513", "end": "2023-05-23 18:46:55.424752", "msg": "non-zero return code", "rc": 2, "start": "2023-05-23 18:46:55.015239", "stderr": "Extension ?securedrop@securedrop.freedom.press? does not exist", "stderr_lines": ["Extension ?securedrop@securedrop.freedom.press? does not exist"], "stdout": "", "stdout_lines": []}

NO MORE HOSTS LEFT *************************************************************

NO MORE HOSTS LEFT *************************************************************

PLAY RECAP *********************************************************************
localhost                  : ok=56   changed=12   unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

ERROR (run with -v for more): Command '['/home/amnesia/Persistent/securedrop/install_files/ansible-base/securedrop-tails.yml', '--ask-become-pass', '-i', '/dev/null']' returned non-zero exit status 2.
```

After reboot per <https://github.com/freedomofpress/securedrop/pull/6712#pullrequestreview-1250724753>:

* [x] `securedrop-admin --force tailsconfig` succeeds.
* [ ] The SecureDrop extension is available.

After another reboot:
    
* [x] The network hook succeeds.
* [ ] The SecureDrop extension is available.
* [x] `securedrop-admin --force tailsconfig` succeeds.
* [ ] The SecureDrop extension is available.

nathandyer commented:
I appreciate the review @cfm! I'm planning to work on this later this afternoon. When it works as intended, it definitely _should_ enable the extension immediately without needing a reboot. That, and the fact it's not launching for you in 5.13, makes me think I might need to tweak a few things for the updated version - when last I tested this branch, those _were_ working as I expected. I'll poke at this and update here once it's behaving properly on my side and ready for another review.

nathandyer committed:
Add support for gnome-shell 3.38 and fix translations.

This explicitly adds support for GNOME Shell 3.38,
and corrects some issues with the gettext system
to make translations work correctly.

It also wraps the user-facing strings in gettext
calls for said translations.

nathandyer commented:
@cfm I just pushed a small change that should allow the extension to work again on gnome-shell 3.38, within the new Wayland-based session in Tails 5.13.

For reasons I have yet to nail down, the system still isn't detecting that it's there during install (just after a reboot). Will continue to investigate that tomorrow morning.

cfm commented on install_files/ansible-base/roles/tails-config/templates/extension.js.in:
```diff
@@ -91,7 +91,7 @@ class Indicator extends PanelMenu.Button {
         });
 
-        let filebrowser = new PopupMenu.PopupMenuItem(_('Open File Browser'));
+        let filebrowser = new PopupMenu.PopupMenuItem('Open File Browser');
```
Why wouldn't we mark this string for translation?

nathandyer replied:
We definitely would, although for some odd reason it was originally the _only_ string marked for translations, so I made that tweak to get everything to a consistent baseline.

I figured out why it was breaking things, and pushed a new commit this morning which should fix the translations and make all the strings properly translatable.

cfm reviewed (commented):
Yes, with 728caa2da4d3931157ece5c9e79199668a9bfe7f I now get the expected results as of <https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1561835528>:

* [x] `securedrop-admin --force tailsconfig` succeeds.
* [ ] The SecureDrop extension is available.

After reboot:
    
* [x] The network hook succeeds.
* [x] The SecureDrop extension is available.
* [x] All of the extension's menu items do what they say.

...and it's *very* nice, @nathandyer!  This really makes it feel like a "SecureDrop admin workstation", rather than a "Tails workstation that happens to have some SecureDrop icons on the desktop".  :-)

nathandyer commented:
I've been working all morning trying to get it to enable the extension on the running system, with no luck. Every piece of documentation I can find is gaslighting me into thinking that it was never possible to enable the extension before reloading gnome-shell, but I know for a fact that this worked when I last tested it before the recent Tails updates.

To give some more context to the issue here, gnome-shell loads extensions during initialization; once it's up and running, if another extension is added, the `gnome-extensions list` utility won't even see it until the session gets restarted.

Before the transition to Wayland in the most recent Tails releases, you could restart the gnome-shell session dynamically by pressing Alt + F2, then typing 'r' and 'Enter'. While I know I _have_ done that on many occasions, I certainly wasn't doing that each time I was running `tailsconfig`. In any case, whether it was working before or not, it's not working now in the Wayland session.

Unfortunately, there isn't a way to dynamically reload gnome-shell in Wayland like there was on X11. The Alt + F2 + r trick is no longer an option. And it steadfastly refuses to acknowledge the newly installed extension until reboot.

I've tried several workarounds this morning to get it to work in-place:

* I tried getting bundling the extension into a .zip and then using the `gnome-extensions install` option to install it. This results in the same state of the extension not being available to enable until next reboot.
* I tried restarting the display manager (gdm). This takes you back to the Tails welcome screen, then you get to a broken state where it never reloads the session and you have to manually power off.
* I tried using `killall gnome-shell` to kill the process, thinking it might automatically restart. It does not (it just dies and gets stuck on a black screen, requiring a manual power off)
* I tried performing a gdbus call like is used for installing a remotely hosted extension. Oddly enough, gnome-shell has the ability to install and activate an extension hosted on extensions.gnome.org within the same session, but this dbus-based call does not work for any local extension, or any extension on any other website.

With that being the case, I don't believe it's going to be possible any longer to enable the extension during the tailsconfig step itself, prior to a reboot. With that being the case, I believe we have three options:

1. We install the extension and prompt the user to reboot after tailsconfig, where it will then load reliably
2. We make the extension load the values dynamically (rather than getting filled in as an Ansible template) and upload it to extensions.gnome.org so that we can use the dbus workaround to activate it at `tailsconfig` time
3. We table the extension indefinitely, until gnome-shell can either activate local extensions immediately on install or until the Wayland-based session can be restarted in-place

nathandyer commented:
After a brief discussion this morning, we have decided to move forward with the first option (prompting the user for a reboot). I will prepare an additional commit that:

1) Removes the logic to install to the running system, instead leaving only the dotfile targets
2) Removes the "enable" step in the ansible task
3) Prompts the user for a reboot

I will also open a follow-up ticket so we can check with the upstream GNOME team to see if it's an intentional choice to have the dbus call only work with extensions hosted on extensions.gnome.org

nathandyer committed:
Prompt user for a reboot and remove logic to install to running system

nathandyer commented:
@cfm Just pinging to let you know that the new round of changes are in, and that this should be in a reviewable state again.

cfm reviewed (dismissed):
I've improvised a compound test plan here to cover as many cases as possible.


## Remove my prior installation of the extension
    
```sh-session
amnesia@amnesia:~$ sudo ls -al /live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions
[sudo] password for amnesia:          
total 12
drwxr-xr-x 3 amnesia amnesia 4096 May 23 18:46 .
drwxr-xr-x 3 amnesia amnesia 4096 May 23 18:46 ..
drwxr-xr-x 4 amnesia amnesia 4096 May 24 22:20 securedrop@securedrop.freedom.press
amnesia@amnesia:~$ rm -rf /live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press
```

* Reboot.
* [x] The SecureDrop extension is gone.


## Reinstall the extension from ccdcccaf75c0bd6162778b77a156a67042c204d0

* [x] `securedrop-admin --force tailsconfig` succeeds.

```sh-session
TASK [tails-config : Remind user to reboot] ************************************
ok: [localhost] => {
    "msg": "Please reboot your Workstation to take advantage of the SecureDrop Extension."
}

TASK [Configuration complete.] *************************************************
ok: [localhost] => {
    "msg": "Successfully configured Tor and set up desktop bookmarks for SecureDrop! You will see a notification appear on your screen when Tor is ready.\nThe Journalist Interface's Tor onion URL is: http://77aml4ggknlndjnkshytyfstwuywlpuncmvr6qnxyuszmzkocobsi4ad.onion The Source Interfaces's Tor onion URL is: http://fu4cp2ynauxf36fdtqluyis75gn33xs72p5bdfu7ve2kut4y7gbqvxid.onion  SSH aliases are set up. You can use them with 'ssh app' and 'ssh mon'. "
}

PLAY RECAP *********************************************************************
localhost                  : ok=61   changed=4    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0  
```

**Feedback:**  Ideally we'd issue just one post-configuration message that addresses both the desktop bookmarks and the extension.  However, "extension" describes the implementation, not the new widget users should look for.  Can we describe the latter as (e.g.) "the SecureDrop menu", with the goal of minimizing the need for documentation to accompany this new feature, even keeping the desktop bookmarks in place as well?


## Test the Admin Workstation experience

* Reboot.
* [x] The network hook succeeds.
* [x] The SecureDrop extension is available.
* [x] All of the extension's menu items are available and do what they say.

## Test the Journalist Workstation experience:

* Move `~/Persistent/securedrop/install_files/ansible-base/tor_v3_keys.json` out of the way.
* Reboot.
* [x] The network hook succeeds.
* [x] The SecureDrop extension is available.
* [x] All of the extension's **non-admin** menu items are available and do what they say.


## Test translations

```sh-session
amnesia@amnesia:~$ sudo tee -a /live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/po/es_ES/LC_MESSAGES/messages.po <<EOF
> msgid "Launch Source Interface"
> msgstr "[test Spanish translation]"
> EOF
```

* Reboot into Spanish (selected at the Tails greeter).
* [ ] The translated string appears in the extension's menu.

**Alas, no.**  After much tinkering, I was able to force this in each of two manual ways:

```sh-session
# First method via gnome-extensions tool:
$ sudo apt install gettext
$ cd /live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press
$ cp po/es_ES/LC_MESSAGES/messages.po po/
$ gnome-extensions pack --gettext-domain securedrop
```

```sh-session
# Second method directly via msgfmt:
$ sudo apt install gettext
$ cd /live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press
$ mkdir -p locale/es_ES/LC_MESSAGES
$ msgfmt po/es_ES/LC_MESSAGES/messages.po -o locale/es_ES/LC_MESSAGES/securedrop.po
```

A literal tl;dr:  *I* didn't read the section ["Packing an Extension with Translations"](https://gjs.guide/extensions/development/translations.html#packing-an-extension-with-translations), because we're not packing this extension.  But it turns out to contain the crucial hint (my emphasis):

> Using the `gnome-extensions` tool makes it easy to **compile and** include the translations with your extension.

This means I was wrong way back in <https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1412619641>: GNOME JavaScript's `Gettext` doesn't have any magic to read `.po`s directly, and we do need to compile them to `.mo`s after all.  (It was too good to be true!)

We have two options:
    
1. If we (CC: @zenmonkeykstop) can live with installing `msgfmt` on Admin and Journalist Workstations via the `gettext` package, then the following patch takes care of everything:

    <details><summary>Install <code>gettext</code>, compile and install catalogs, and correct gettext domain</summary>

    ```patch
    diff --git a/admin/bootstrap.py b/admin/bootstrap.py
    index 25974d949..26ac0147a 100755
    --- a/admin/bootstrap.py
    +++ b/admin/bootstrap.py
    @@ -119,6 +119,7 @@ def install_apt_dependencies(args: argparse.Namespace) -> None:
             "-c",
             "apt-get update && \
                        apt-get -q -o=Dpkg::Use-Pty=0 install -y \
    +                   gettext \
                        python3-virtualenv \
                        python3-yaml \
                        python3-pip \
    diff --git a/install_files/ansible-base/roles/tails-config/tasks/install_shell_extension.yml b/install_files/ansible-base/roles/tails-config/tasks/install_shell_extension.yml
    index 2931a7d50..23adbc42c 100644
    --- a/install_files/ansible-base/roles/tails-config/tasks/install_shell_extension.yml
    +++ b/install_files/ansible-base/roles/tails-config/tasks/install_shell_extension.yml
    @@ -117,7 +117,11 @@
         - "{{ _securedrop_extension_info }}"
         - "{{ tails_config_extension_directories }}"
     
    +- name: Compile translations
    +  command: |
    +    find {{ tails_config_amnesia_persistent }}/securedrop/securedrop/translations -name '*.po' -execdir msgfmt \{\} \;
    +
     - name: Add extension translations in Persistent Storage
       synchronize:
         src: "{{ tails_config_amnesia_persistent }}/securedrop/securedrop/translations/"
    -    dest: "{{ tails_config_live_dotfiles }}/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/po/"
    +    dest: "{{ tails_config_live_dotfiles }}/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/locale/"
    diff --git a/install_files/ansible-base/roles/tails-config/templates/extension.js.in b/install_files/ansible-base/roles/tails-config/templates/extension.js.in
    index cd2296cda..91253dd42 100644
    --- a/install_files/ansible-base/roles/tails-config/templates/extension.js.in
    +++ b/install_files/ansible-base/roles/tails-config/templates/extension.js.in
    @@ -1,6 +1,6 @@
     /* exported init */
     
    -const GETTEXT_DOMAIN = 'securedrop';
    +const GETTEXT_DOMAIN = 'messages';
     
     const {Clutter, GObject, GLib } = imports.gi;

    ```
    </details>

2. Otherwise, we'll need to go back to our original plan of folding these strings into the `securedrop-desktop` Weblate component after all, via `i18n_tool.py --translate-desktop` (at least until continuous translation lets us simplify this tooling).

I'd appreciate both of your reactions, @nathandyer and @zenmonkeykstop.  I realize (2) is a bit hand-wavy; I'm happy to tackle this next week if we decide to go this route.  It won't be a big commit to tack on, just arcane to implement and byzantine to test.

nathandyer commented:
Thank you so much, @cfm! I'll make the change so that we only see the one post-install message, and that it refers to "the SecureDrop menu" rather than "extension" (I think that's a very reasonable and user-friendly way of describing that menu).

As for the options with translations, either option sounds good to me. If @zenmonkeykstop is agreeable to option #1, that feels like the cleanest option. If not, though, I'm happy to work through option #2 (a lot of the work re: i18n_tool.py was already done at an earlier step anyway) with you this week.

nathandyer committed:
Only show one message to the user at end of the run, and refer to the extension as the SecureDrop Menu.

cfm commented:
Discussed with @nathandyer and @zenmonkeykstop:  I'll reconstruct option (2) in <https://github.com/freedomofpress/securedrop/pull/6712#pullrequestreview-1444812440> (now gone from GitHub's cache of commits no longer part of this branch), smoke-test on my end, and return to you two with a localization-specific test plan for tomorrow.  For simplicity, this keeps the GNOME Shell extension localized in the same way as the desktop shortcuts, and then both will move together to the new unified workflow for continuous localization.

cfm committed:
build(gettext): extend the "translate-desktop" command to cover the GNOME Shell extension

In the original (since lost) 1cfd35b, we tried to treat the GNOME Shell
extension as a separate "securedrop/extension" Weblate component with
its own "translate-extension" command, replicating both
"--extract-update" and "--compile" modes.

In f3d3f04, we tried to treat the GNOME Shell extension as an extension
(sorry) of the main "securedrop/securedrop" component, taking advantage
of Babel's mapping feature to parse the ".js.in" file as JavaScript.

Here we harmonize the two approaches.  "translate-desktop
--extract-update" is now a two-step process, first via Babel over the
".js.in" JavaScript file, then as usual via xgettext over the ".j2.in"
desktop templates.  (This is logically reversed, but Babel appears to
have no equivalent of xgettext's appending "--join-existing" mode.)
"translate-desktop --compile" similarly does two passes with msgfmt,
first to create the standard gettext directory layout and then to update
the desktop templates.

NB. We also add "msgmerge --no-fuzzy-matching" after #6772, which
required rebasing from "develop".

cfm committed:
deploy(tails-config): expect "messages" domain in "locale" directory

In Ansible idiom, the "locale" tree should come from the role's files
rather than its template, which would simplify the path reference.
However, it's generated by "i18n_tool.py", which shouldn't be concerned
with the surrounding Ansible layout.

cfm commented:
<https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1569078315>:

> Discussed with @nathandyer and @zenmonkeykstop: I'll reconstruct option (2) in [#6712 (review)](https://github.com/freedomofpress/securedrop/pull/6712#pullrequestreview-1444812440) (now gone from GitHub's cache of commits no longer part of this branch), smoke-test on my end, and return to you two with a localization-specific test plan for tomorrow. For simplicity, this keeps the GNOME Shell extension localized in the same way as the desktop shortcuts, and then both will move together to the new unified workflow for continuous localization.

This is done, and 76f3adeed90f4aaadbf0685e09dec6314367d5c0 has the gory implementation details.  Procedurally:
1. This required rebasing from `develop` for #6772.
2. The diff to `i18n_tool.py` is the frankenchore I had hoped to avoid, but it manages to accommodate both the desktop shortcuts and the extension without introducing any new components in Weblate *or* any new steps in the current localization workflow.  I very much look forward to ripping it out along with all the rest of `i18n_tool.py` per freedomofpress/securedrop-engineering#30.
3. Please rebase and squash!

**To test `i18n_tool.py translate-desktop` with the GNOME Shell extension,** on this branch:

1. `securedrop/bin/dev-shell ./i18n_tool.py translate-desktop --extract-update`
2. Check `desktop.pot` and one `.po`.

    <details><summary>e.g.</summary>

    ```sh-session
    (.venv) user@sd-dev:~/securedrop$ git diff *.pot **/es_ES.po
    diff --git a/install_files/ansible-base/roles/tails-config/templates/desktop.pot b/install_files/ansible-base/roles/tails-config/templates/desktop.pot
    index 524987d44..c1ee0f43f 100644
    --- a/install_files/ansible-base/roles/tails-config/templates/desktop.pot
    +++ b/install_files/ansible-base/roles/tails-config/templates/desktop.pot
    @@ -13,9 +13,30 @@ msgstr ""
     "Language-Team: LANGUAGE <LL@li.org>\n"
     "Language: \n"
     "MIME-Version: 1.0\n"
    -"Content-Type: text/plain; charset=CHARSET\n"
    +"Content-Type: text/plain; charset=utf-8\n"
     "Content-Transfer-Encoding: 8bit\n"
     
    +msgid "Launch Source Interface"
    +msgstr ""
    +
    +msgid "Launch Journalist Interface"
    +msgstr ""
    +
    +msgid "Check for SecureDrop Updates"
    +msgstr ""
    +
    +msgid "SSH into the App Server"
    +msgstr ""
    +
    +msgid "SSH into the Monitor Server"
    +msgstr ""
    +
    +msgid "Open KeePassXC Password Vault"
    +msgstr ""
    +
    +msgid "Open File Browser"
    +msgstr ""
    +
     #: desktop-journalist-icon.j2.in:10
     msgid "SecureDrop Journalist Interface"
     msgstr ""
    diff --git a/install_files/ansible-base/roles/tails-config/templates/es_ES.po b/install_files/ansible-base/roles/tails-config/templates/es_ES.po
    index ccb841eb2..accdb186a 100644
    --- a/install_files/ansible-base/roles/tails-config/templates/es_ES.po
    +++ b/install_files/ansible-base/roles/tails-config/templates/es_ES.po
    @@ -18,6 +18,27 @@ msgstr ""
     "Plural-Forms: nplurals=2; plural=n != 1;\n"
     "X-Generator: Weblate 2.20\n"
     
    +msgid "Launch Source Interface"
    +msgstr ""
    +
    +msgid "Launch Journalist Interface"
    +msgstr ""
    +
    +msgid "Check for SecureDrop Updates"
    +msgstr ""
    +
    +msgid "SSH into the App Server"
    +msgstr ""
    +
    +msgid "SSH into the Monitor Server"
    +msgstr ""
    +
    +msgid "Open KeePassXC Password Vault"
    +msgstr ""
    +
    +msgid "Open File Browser"
    +msgstr ""
    +
     #: desktop-journalist-icon.j2.in:10
     msgid "SecureDrop Journalist Interface"
     msgstr "Interfaz de periodista de SecureDrop"
    ```
    </details>

3. Edit `es_ES.po` so that `Launch Source Interface` has a `msgstr` of `[test translation]`.
4. `securedrop/bin/dev-shell ./i18n_tool.py translate-desktop --compile`
5. Commit *all* of `install_files/ansible-base/roles/tails-config/templates` to a test branch and push it.
6. Check out your test branch on Tails.
7. Install the extension with `./securedrop-admin --force tailsconfig`.
8. Reboot into Spanish (selected at the Tails greeter).
9. [ ] The translated string appears in the extension's menu.
10. [ ] All of the extension's menu items are available and do what they say.

nathandyer commented:
Thanks @cfm! I just stepped through the test plan and can confirm everything is working as expected!

- [x] The translated string appears in the extension's menu.
- [x] All of the extension's menu items are available and do what they say.

![Captura de pantalla de 2023-05-31 18-52-26](https://github.com/freedomofpress/securedrop/assets/6131358/63ce7c79-9840-45b4-9b1c-ce463c380a5d)

cfm commented:
Hooray; thanks, @nathandyer!  All yours, @zenmonkeykstop.

zenmonkeykstop commented on install_files/ansible-base/roles/tails-config/tasks/install_shell_extension.yml:
```diff
@@ -0,0 +1,123 @@
+    - "{{ _securedrop_extension_info }}"
+    - "{{ tails_config_extension_directories }}"
+
+- name: Add extension translations in Persistent Storage
```
On a fresh install, `./securedrop-admin --force tailsconfig` is failing for me here as follows:
```
TASK [tails-config : Add extension translations in Persistent Storage] *********
fatal: [localhost]: FAILED! => {"changed": false, "cmd": "/usr/bin/rsync --delay-updates -F --compress --archive --out-format=<<CHANGED>>%i %n%L /home/amnesia/Persistent/securedrop/install_files/ansible-base/roles/tails-config/templates/locale/ /live/persistence/TailsData_unlocked/dotfiles/.local/share/gnome-shell/extensions/securedrop@securedrop.freedom.press/locale/", "msg": "rsync: [sender] change_dir \"/home/amnesia/Persistent/securedrop/install_files/ansible-base/roles/tails-config/templates/locale\" failed: No such file or directory (2)\nrsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1333) [sender=3.2.3]\n", "rc": 23}
```

cfm replied:
Clarified out of band:  This can be bootstrapped from <https://github.com/freedomofpress/securedrop/pull/6712#issuecomment-1569440679>, skipping step (3), and committing those changes onto *this* branch rather than the localization-specific testing branch I suggested there.

zenmonkeykstop replied:
Confirmed, fresh install working with change above.

zenmonkeykstop reviewed (changes requested):
Fresh install failing, further review a bit blocked

zenmonkeykstop committed:
Adding placeholder locale files

zenmonkeykstop reviewed (approved):
Tested the following (on a virtualized setup): 

### Fresh install, ssh over local
- [x] Fresh install using this branch on the Admin Workstation completes successfully
- [x]  Menu is available after rebooting, with icon displayed correctly
- [x] Journalist (interfaces, updater, file browser) and Admin (SSH to server) menu options are displayed and work as expected

### Admin Upgrade (any ssh option)
- set up a prod (VM OK) install using the current latest version (2.5.2)
- On the Admin workstation, simulate running the GUI updater via these commands in sequence:
  ```
  cd ~/amnesia/Persistent/securedrop && git checkout gnome-shell-extension
  ./securedrop-admin setup
  ./securedrop-admin --force tailsconfig
  ```
- [x] `tailsconfig` completes successfully, prompting user to reboot
- [x]  Menu is available after rebooting, with icon displayed correctly
- [x] Journalist (interfaces, updater, file browser) and Admin (SSH to server) menu options are displayed and work as expected


### Journalist workstation - fresh install
- [x] Following JW setup instructions and rebooting results in menu being available
- [x] Admin (ssh) options are *not* displayed, other options are and work as expected

### Journalist workstation - upgrade 
- On an existing Journalist workstation, set up with current release (2.5.2), simulate running the GUI updater via these commands in sequence:
  ```
  cd ~/amnesia/Persistent/securedrop && git checkout gnome-shell-extension
  ./securedrop-admin setup
  ./securedrop-admin --force tailsconfig
  ```
- [x] `tailsconfig` completes successfully, prompting user to reboot
- [x]  Menu is available after rebooting, with icon displayed correctly
- [x] Journalist (interfaces, updater, file browser)  menu options are displayed and work as expected
- [x] Admin (ssh) options are *not dispalyed

Visual review also good, LGTM!
